receiver/namedpipereceiver/namedpipe_test.go

```diff
@@ -63,6 +63,7 @@ func TestReadPipe(t *testing.T) {
 	rcvr, err := f.CreateLogsReceiver(context.Background(), receivertest.NewNopCreateSettings(), cfg, sink)
 	require.NoError(t, err, "failed to create receiver")
 	require.NoError(t, rcvr.Start(context.Background(), componenttest.NewNopHost()))
+	defer require.NoError(t, rcvr.Shutdown(context.Background()))
```

Comment:
I don't know if this fixes a leak or not, but it should be included and doesn't hurt anything.

Reply:
Looks like this caused the test to timeout:
```
panic: test timed out after 5m0s
```
The stanza log receiver expects the context to be cancelled to properly return, so I'll modify the test to call cancel on the context.

Reply:
I'll continue sharing debugging here. The stack traces collected from the time out panic show the main threads are blocked in `pkg/stanza/converter` loops (which is expected if shutdown hasn't been called yet), and a goroutine started by the test opening a file:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9124bc85cabfbc5bd20d3046b5d2248c69dffb2e/receiver/namedpipereceiver/namedpipe_test.go#L67

From a quick search, it looks like this is [expected behavior when opening a namedpipe](https://github.com/golang/go/issues/33050) with only read or write (not both) permissions. This leads me to wonder why this was passing before my changes here.

_Still investigating. I only have access to a Mac so I've had to test using GitHub's actions here._

Reply:
The test is back to passing again without calling shutdown, I'm not sure why. `goleak` passes without calling shutdown, so obviously the context is being handled properly internally with cancel being called.

I'm going to check if stopping the converter first will allow the receiver to shutdown properly.

Reply:
Turns out I was using `defer` incorrectly. To pick up the right context, the `require` call has to be done in a `func` closure, this way the arguments to `require` won't be evaluated and set until the `defer` itself is executed.

[Stack Overflow reference](https://stackoverflow.com/questions/47321012/difference-between-defer-call-a-function-directly-and-call-it-in-an-anonymous-fu)
[Documentation reference](https://go.dev/blog/defer-panic-and-recover)